### PR TITLE
DAO8

### PR DESCRIPTION
**Pull request by Max-Pochepets**
DataBase web description complete

### CODE REVIEW

File: src/main/webapp/WEB-INF/views/manufacturers/create.jsp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    <head>
+            <title>Create manufacturer</title>
+        </head>
+        <body>
+        <h1>Please provide manufacturers credentials.</h1>
+
+        <form method="post" action="${pageContext.request.contextPath}/manufacturers/create">
+            Manufacturer's name:<input type="text" name="name">
+            Manufacturer's country:<input type="text" name="country">
+
```
Comment (semyonich) [inline src/main/webapp/WEB-INF/views/manufacturers/create.jsp:12]: IMHO you should check input data for non empty value

File: src/main/webapp/WEB-INF/views/drivers/create.jsp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+<html>
+    <head>
+        <title>Create driver</title>
+    </head>
+    <body>
+        <h1>Please provide driver's credentials.</h1>
+
+        <form method="post" action="${pageContext.request.contextPath}/drivers/create">
+            Driver's name:<input type="text" name="name">
+            Driver's license number:<input type="text" name="license">
```
Comment (semyonich) [inline src/main/webapp/WEB-INF/views/drivers/create.jsp]: same

File: src/main/webapp/WEB-INF/views/cars/drivers/delete.jsp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Delete driver from car</title>
+</head>
+<body>
+    <form method="post" action="${pageContext.request.contextPath}/cars/drivers/delete">
+        Car's id:<input type="text" name="carID">
+        Driver's id:<input type="text" name="driverID">
+
```
Comment (semyonich) [inline src/main/webapp/WEB-INF/views/cars/drivers/delete.jsp:12]: what if carId will be "test" ? check input

File: src/main/java/crud/controller/driver/GetAllDriversController.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import java.io.IOException;
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class GetAllDriversController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final DriverService driverService
```
Comment (semyonich) [inline src/main/java/crud/controller/driver/GetAllDriversController.java]: it's not a CONSTANT, use common camelCase

File: src/main/java/crud/controller/driver/DeleteDriversController.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import crud.service.abstraction.DriverService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class DeleteDriversController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final DriverService driverService
```
Comment (semyonich) [inline src/main/java/crud/controller/driver/DeleteDriversController.java]: same

File: src/main/java/crud/controller/driver/CreateDriversController.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import crud.service.abstraction.DriverService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class CreateDriversController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final DriverService driverService
```
Comment (semyonich) [inline src/main/java/crud/controller/driver/CreateDriversController.java]: check it everywhere

**Review by semyonich — CHANGES_REQUESTED** [review]

File: src/main/java/crud/controller/driver/GetAllDriversController.java (hunk 4 above)
Comment (Max-Pochepets) [reply src/main/java/crud/controller/driver/GetAllDriversController.java]: maybe I've misunderstood, but why is it not?

File: src/main/webapp/WEB-INF/views/drivers/create.jsp (hunk 2 above)
Comment (semyonich) [inline src/main/webapp/WEB-INF/views/drivers/create.jsp]: check input

File: src/main/webapp/WEB-INF/views/manufacturers/create.jsp (hunk 1 above)
Comment (semyonich) [inline src/main/webapp/WEB-INF/views/manufacturers/create.jsp:12]: check input

**Review by semyonich — APPROVED** [review]

File: src/main/java/crud/controller/ClearDataController.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class ClearDataController extends HttpServlet {
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException {
+        ConnectionUtil.clearTable();
+        resp.sendRedirect(req.getContextPath() + "/");
+    }
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/ClearDataController.java]: I don't think that you should have such functionality in your project
Imagine : https://rozetka.com.ua/ua/computers-notebooks/clear

File: src/main/java/crud/controller/InjectDataController.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            = new Driver("John", "AAAEWWW222JHDSF8899");
+    private static final Driver DRIVER_BOB
+            = new Driver("BOB", "IEH709898SDF7689SDF231");
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final ManufacturerService MANUFACTURER_SERVICE
+            = (ManufacturerService) INJECTOR.getInstance(ManufacturerService.class);
+    private static final CarService CAR_SERVICE
+            = (CarService) INJECTOR.getInstance(CarService.class);
+    private static final DriverService DRIVER_SERVICE
+            = (DriverService) INJECTOR.getInstance(DriverService.class);
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/InjectDataController.java]: nither of those are constants

File: src/main/java/crud/controller/car/AddDriverToCarController.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AddDriverToCarController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final CarService CAR_SERVICE
+            = (CarService) INJECTOR.getInstance(CarService.class);
+    private static final DriverService DRIVER_SERVICE
+            = (DriverService) INJECTOR.getInstance(DriverService.class);
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/car/AddDriverToCarController.java]: same

File: src/main/java/crud/controller/car/CreateCarsController.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class CreateCarsController extends HttpServlet {
+    private static final Injector INJECTOR = Injector.getInstance("crud");
+    private static final CarService CAR_SERVICE
+            = (CarService) INJECTOR.getInstance(CarService.class);
+    private static final ManufacturerService MANUFACTURER_SERVICE
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/car/CreateCarsController.java]: same

File: src/main/java/crud/controller/car/DeleteDriverFromCar.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/cars/drivers/delete.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        try {
+            String carID = req.getParameter("carID");
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/car/DeleteDriverFromCar.java]:
```suggestion
            String carID = req.getParameter("car_id");
```
Change all parameters name in such manner

File: src/main/java/crud/controller/driver/CreateDriversController.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import crud.lib.Injector;
+import crud.model.Driver;
+import crud.service.abstraction.DriverService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class CreateDriversController extends HttpServlet {
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/driver/CreateDriversController.java]:
```suggestion
public class CreateDriverController extends HttpServlet {
```

File: src/main/java/crud/controller/driver/CreateDriversController.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/drivers/create.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException {
+        String name = req.getParameter("name");
+        String license = req.getParameter("license");
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/driver/CreateDriversController.java]:
```suggestion
        String name = req.getParameter("driver_name");
        String license = req.getParameter("driver_license");
```

File: src/main/java/crud/controller/driver/DeleteDriversController.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+import crud.lib.Injector;
+import crud.service.abstraction.DriverService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class DeleteDriversController extends HttpServlet {
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/driver/DeleteDriversController.java]:
```suggestion
public class DeleteDriverController extends HttpServlet {
```

File: src/main/java/crud/controller/manufacturer/DeleteManufacturersController.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+import crud.lib.Injector;
+import crud.service.abstraction.ManufacturerService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class DeleteManufacturersController extends HttpServlet {
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/manufacturer/DeleteManufacturersController.java]:
```suggestion
public class DeleteManufacturerController extends HttpServlet {
```

File: src/main/webapp/WEB-INF/views/cars/all.jsp [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+                            <c:out value="${driver.name}"/>
+                            <br />
+                        </c:forEach>
+                    </td>
+                    <td>
+                        <c:forEach var="driver" items="${car.drivers}">
+                            <c:out value="${driver.licenseNumber}"/>
+                            <br />
+                        </c:forEach>
+                    </td>
```
Comment (dmytro-hryhoruk) [inline src/main/webapp/WEB-INF/views/cars/all.jsp:44]: could you do this with one loop?

File: src/main/java/crud/controller/car/DeleteDriverFromCar.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+            String driverID = req.getParameter("driverID");
+            Car car = CAR_SERVICE.get(Long.valueOf(carID));
+            Driver driver = driverService.get(Long.valueOf(driverID));
+            CAR_SERVICE.removeDriverFromCar(driver, car);
+            resp.sendRedirect(req.getContextPath() + "/cars/");
+        } catch (NoSuchElementException e) {
+            req.setAttribute("message", "Invalid data. Please check if car/driver exists.");
+            req.getRequestDispatcher("/WEB-INF/views/cars/drivers/delete.jsp").forward(req, resp);
```
Comment (dmytro-hryhoruk) [inline src/main/java/crud/controller/car/DeleteDriverFromCar.java]: let's not catch runtime exceptions. In the scope of this task we can just believe that we will get the real id's from the frontend

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/main/webapp/WEB-INF/views/cars/all.jsp (hunk 16 above)
Comment (Max-Pochepets) [reply src/main/webapp/WEB-INF/views/cars/all.jsp:44]: it will look awful

File: src/main/java/crud/controller/ClearDataController.java (hunk 7 above)
Comment (Max-Pochepets) [reply src/main/java/crud/controller/ClearDataController.java]: for the purpose of testing

File: src/main/java/crud/controller/car/AddDriverToCarController.java (hunk 9 above)
Comment (Max-Pochepets) [reply src/main/java/crud/controller/car/AddDriverToCarController.java]: why?

Comment (Max-Pochepets) [reply src/main/java/crud/controller/car/AddDriverToCarController.java]: nvm, I got it)

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**